core/server/master/src/main/java/alluxio/master/FaultTolerantAlluxioMasterProcess.java

```diff
@@ -73,15 +73,21 @@ protected FaultTolerantAlluxioMasterProcess(JournalSystem journalSystem,
   public void start() throws Exception {
     mRunning = true;
     mJournalSystem.start();
+
+    startMasters(false);
+    LOG.info("Secondary started");
+
+    if (ServerConfiguration.getBoolean(PropertyKey.MASTER_JOURNAL_INITIAL_REPLAY_ENABLED)) {
+      mJournalSystem.waitForInitialReplay();
```

Comment:
I was wondering if this initial replay will only work for the process start? I originally thought it could be ok to do whenever a master process became secondary, basically doing this wait within the while loop below. What are the possible issues with that?

Reply:
Was it already decided that we will only do the replay for the very first replay, and not other replays?

Reply:
Sorry for the late reply. I am still getting more information on whether I am able to add the initial replay to the while loop safe and whether the change can bring more benefits. There hasn't been any decision but I first move it out to make sure the behavior is understandable. 

In the current implementation, the initial replay is put before we connect to Zookeeper and join leader election to make sure the initial replay can finish and not affect the other running Alluxio masters.  In the previous implementation which we put the `waitForInitReplay` only in the while loop, the following behavior may happen:
```
For example, 
* Master A is the leading master
* Master B join the leader election
* Master B inits from replay
* During the Master B initial replay, Master B is selected as the leading master
* If Master B replaying for too long, it cannot gain primacy and able to serve requests fast enough, then no Alluxio Master will be able to serve client requests in a long period.
```
As for whether we want to add `waitForCatchup` in the while loop, it may not make a huge difference if all masters already finish the initial replay. Step 3 is not expected to take a long time.
```
1. mJournalSystem.waitForInitialReplay();
2. mLeaderSelector.start(getRpcAddress());
 while (!Thread.interrupted()) {
3. mJournalSystem.waitForCatchup();   // I think this can be added may not make a difference
      mLeaderSelector.waitForState(State.PRIMARY);
      if (!mRunning) {
        break;
      }
      if (gainPrimacy()) {
        mLeaderSelector.waitForState(State.SECONDARY);
        if (!mRunning) {
          break;
        }
        losePrimacy();
      }
```

Reply:
@gpang PTAL, thanks!

Reply:
Yes, if the `waitForCatchup` is in the while loop, the behavior you described would happen. So, the worst case is that for some period of time, there will be no master service requests. However, is that different from any alternative? For example, the leader will only serve after it finishes `gainPrimacy()`.

Isn't the `waitForState()` method the only place where a master will consult the zk state? So, as long as we `waitForCatchup()` before `waitForState()`, we will always finish replaying the journal before checking the zk state.

Reply:
Yeah, there should be no difference between catchup and then gainPrimacy or gainPrimacy different when considering no master service period.
Adding waitForCatchup can also make sure the master catches up during frequent leadership change.
Overall, adding waitForCatchup in the while loop sounds to be slightly better than the current implementation. 
@bf8086 What's your opinion about adding the waitForCatchup in the Primary/Seconday while loop
```
 while (!Thread.interrupted()) {
     mJournalSystem.waitForCatchup();  <--- adding this
      mLeaderSelector.waitForState(State.PRIMARY);
      if (!mRunning) {
        break;
      }
      if (gainPrimacy()) {
        mLeaderSelector.waitForState(State.SECONDARY);
        if (!mRunning) {
          break;
        }
        losePrimacy();
      }
```

Reply:
Agree it's helpful to add `waitForCatchup` in the while loop for frequent leadership changes. Can you maybe try some failover tests to make sure it does not introduce undesired behaviors?

Reply:
Yeah, my intuition feels like it should be correct in the while-loop, but I'm just not completely sure. Running some tests would increase our confidence.